### PR TITLE
feat: add kubeone plugin

### PR DESCRIPTION
**Pull request by nerdeveloper**

### CODE REVIEW

File: plugins/kubeone.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    - selector:
+        matchLabels:
+          os: darwin
+          arch: amd64
+      uri: https://github.com/kubermatic/kubeone/releases/download/v1.3.0/kubeone_1.3.0_darwin_amd64.zip
+      sha256: 3dc786156c48acfabb7f9ae39483ec2ea12983bb371d0f447095d5a0afeca96b
+      bin: kubeone
+  shortDescription: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
+  homepage: https://github.com/kubermatic/kubeone
+  caveats: ""
```
Comment (corneliusweig) [inline plugins/kubeone.yaml:24]: Please drop that field if there is no caveat.

**Review by corneliusweig — COMMENTED** [review]
You also need to sign the CLA before we can move forward.

File: plugins/kubeone.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      sha256: 810cee80f8adcbc90b836525b70869641e0e4f079a3dda05dd3187b708c38248
+      bin: kubeone
+    - selector:
+        matchLabels:
+          os: darwin
+          arch: amd64
+      uri: https://github.com/kubermatic/kubeone/releases/download/v1.3.0/kubeone_1.3.0_darwin_amd64.zip
+      sha256: 3dc786156c48acfabb7f9ae39483ec2ea12983bb371d0f447095d5a0afeca96b
+      bin: kubeone
+  shortDescription: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
```
[the remark this answers is not in the document]
Comment (nerdeveloper) [reply plugins/kubeone.yaml]: Done

File: plugins/kubeone.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  homepage: https://github.com/kubermatic/kubeone
+  caveats: ""
+  description: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
```
[the remark this answers is not in the document]
Comment (nerdeveloper) [reply plugins/kubeone.yaml]: Done

**Comment by nerdeveloper** [comment]
Hey @ahmetb @corneliusweig I am not sure why the pipeline keeps failing, The krew manifest deploys well on my local ubuntu setup

Please, is there a special flag i need to pass to get it installed for mac OS, the zip format does not work? 

Command: `kubectl krew  install --manifest plugins/kubeone.yaml`
Screenshot 2021-11-01 at 10 58 16

File: plugins/kubeone.yaml (hunk 1 above)
Comment (nerdeveloper) [reply plugins/kubeone.yaml]: Done

File: plugins/kubeone.yaml [hunk 4]
```diff
@@ -19,7 +19,10 @@ spec:
...
       bin: kubeone
-  shortDescription: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
+  shortDescription: Automate cluster operations on chosen environments
   homepage: https://github.com/kubermatic/kubeone
-  caveats: ""
-  description: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
+  description: |
+    The offical plugin for Kubeone, a kubernetes tool for
+    automating cluster operations on all your cloud, on-prem, edge,
+    and IoT environments.
```
Comment (corneliusweig) [inline plugins/kubeone.yaml:27]: Can you please elaborate what are the "cluster operations"? Please describe the plugin in a way that someone with a technical background can understand what is going on. The current description sounds too much like marketing.

File: plugins/kubeone.yaml [hunk 5]
```diff
@@ -19,7 +19,10 @@ spec:
       uri: https://github.com/kubermatic/kubeone/releases/download/v1.3.0/kubeone_1.3.0_darwin_amd64.zip
       sha256: 3dc786156c48acfabb7f9ae39483ec2ea12983bb371d0f447095d5a0afeca96b
       bin: kubeone
-  shortDescription: Kubermatic KubeOne automate cluster operations on all your cloud, on-prem, edge, and IoT environments
+  shortDescription: Automate cluster operations on chosen environments
```
Comment (corneliusweig) [inline plugins/kubeone.yaml:22]: Please be more specific what are "cluster operations". I also don't understand why you emphasisze "chosen environments". This seem obvious to me, so there's no need to mention it?

**Review by corneliusweig — COMMENTED** [review]
I don't quite understand why the CI install is failing. Can you please check the installation locally, faking a mac environment, like here: https://github.com/kubernetes-sigs/krew/blob/1649da8245dd4e956301a7940e2e606caa941341/cmd/validate-krew-manifest/main.go#L148-L153

**Comment by nerdeveloper** [comment]
Thank you for the comment

this is the logs
```bash
kubectl krew  install --manifest plugins/kubeone.yaml -v=4
I1101 11:50:10.636399   30576 root.go:221] Ensure creating dir: "/Users/odirionye/.krew"
I1101 11:50:10.636499   30576 root.go:221] Ensure creating dir: "/Users/odirionye/.krew/store"
I1101 11:50:10.636511   30576 root.go:221] Ensure creating dir: "/Users/odirionye/.krew/bin"
I1101 11:50:10.636522   30576 root.go:221] Ensure creating dir: "/Users/odirionye/.krew/index"
I1101 11:50:10.636532   30576 root.go:221] Ensure creating dir: "/Users/odirionye/.krew/receipts"
I1101 11:50:10.636644   30576 root.go:128] skipping upgrade check
I1101 11:50:10.636712   30576 migration.go:30] Checking if index migration is needed.
I1101 11:50:10.636731   30576 migration.go:33] Index already migrated.
I1101 11:50:10.637057   30576 install.go:191] --manifest specified, not ensuring plugin index
I1101 11:50:10.640563   30576 install.go:148] Will install plugin: detached/kubeone
Installing plugin: kubeone
I1101 11:50:10.640582   30576 install.go:58] Looking for installed versions
I1101 11:50:10.640605   30576 platform.go:43] Matching platform for labels(arch=amd64,os=darwin)
I1101 11:50:10.640644   30576 platform.go:51] Found matching platform with index (1)
I1101 11:50:10.640649   30576 install.go:77] Install plugin kubeone at version=v1.3.0
I1101 11:50:10.640656   30576 install.go:94] Creating download staging directory
I1101 11:50:10.640778   30576 install.go:99] Successfully created download staging directory "/var/folders/y3/_y30jhts1ms04mtwjm2swq0c0000gn/T/krew-downloads860795623"
I1101 11:50:10.641148   30576 fetch.go:39] Fetching "https://github.com/kubermatic/kubeone/releases/download/v1.3.0/kubeone_1.3.0_darwin_amd64.zip"
I1101 11:50:12.520162   30576 downloader.go:42] Reading archive file into memory
I1101 11:50:28.085659   30576 downloader.go:47] Read 12833739 bytes from archive into memory
I1101 11:50:28.085697   30576 verifier.go:51] Compare sha256 (3dc786156c48acfabb7f9ae39483ec2ea12983bb371d0f447095d5a0afeca96b) signed version
I1101 11:50:28.085722   30576 downloader.go:203] detected "application/zip" file type
I1101 11:50:28.085732   30576 downloader.go:54] Extracting zip archive to "/var/folders/y3/_y30jhts1ms04mtwjm2swq0c0000gn/T/krew-downloads860795623"
I1101 11:50:28.087741   30576 install.go:101] Deleting the download staging directory /var/folders/y3/_y30jhts1ms04mtwjm2swq0c0000gn/T/krew-downloads860795623
W1101 11:50:28.088517   30576 install.go:164] failed to install plugin "kubeone": install failed: failed to unpack into staging dir: failed to unpack the plugin archive: failed to extract file: can't create file in zip destination dir: open /var/folders/y3/_y30jhts1ms04mtwjm2swq0c0000gn/T/krew-downloads860795623/addons/backups-restic/README.md: no such file or directory
F1101 11:50:28.088558   30576 root.go:77] open /var/folders/y3/_y30jhts1ms04mtwjm2swq0c0000gn/T/krew-downloads860795623/addons/backups-restic/README.md: no such file or directory
can't create file in zip destination dir
sigs.k8s.io/krew/internal/download.extractZIP
	/home/runner/work/krew/krew/internal/download/downloader.go:81
sigs.k8s.io/krew/internal/download.extractArchive
	/home/runner/work/krew/krew/internal/download/downloader.go:208
sigs.k8s.io/krew/internal/download.Downloader.Get
	/home/runner/work/krew/krew/internal/download/downloader.go:233
sigs.k8s.io/krew/internal/installation.downloadAndExtract
	/home/runner/work/krew/krew/internal/installation/install.go:148
sigs.k8s.io/krew/internal/installation.install
	/home/runner/work/krew/krew/internal/installation/install.go:106
sigs.k8s.io/krew/internal/installation.Install
	/home/runner/work/krew/krew/internal/installation/install.go:78
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:156
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
failed to extract file
sigs.k8s.io/krew/internal/download.extractArchive
	/home/runner/work/krew/krew/internal/download/downloader.go:208
sigs.k8s.io/krew/internal/download.Downloader.Get
	/home/runner/work/krew/krew/internal/download/downloader.go:233
sigs.k8s.io/krew/internal/installation.downloadAndExtract
	/home/runner/work/krew/krew/internal/installation/install.go:148
sigs.k8s.io/krew/internal/installation.install
	/home/runner/work/krew/krew/internal/installation/install.go:106
sigs.k8s.io/krew/internal/installation.Install
	/home/runner/work/krew/krew/internal/installation/install.go:78
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:156
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
failed to unpack the plugin archive
sigs.k8s.io/krew/internal/installation.downloadAndExtract
	/home/runner/work/krew/krew/internal/installation/install.go:149
sigs.k8s.io/krew/internal/installation.install
	/home/runner/work/krew/krew/internal/installation/install.go:106
sigs.k8s.io/krew/internal/installation.Install
	/home/runner/work/krew/krew/internal/installation/install.go:78
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:156
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
failed to unpack into staging dir
sigs.k8s.io/krew/internal/installation.install
	/home/runner/work/krew/krew/internal/installation/install.go:107
sigs.k8s.io/krew/internal/installation.Install
	/home/runner/work/krew/krew/internal/installation/install.go:78
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:156
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
install failed
sigs.k8s.io/krew/internal/installation.Install
	/home/runner/work/krew/krew/internal/installation/install.go:85
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:156
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
failed to install some plugins: [kubeone]
sigs.k8s.io/krew/cmd/krew/cmd.init.2.func1
	/home/runner/work/krew/krew/cmd/krew/cmd/install.go:185
github.com/spf13/cobra.(*Command).execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850
github.com/spf13/cobra.(*Command).ExecuteC
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958
github.com/spf13/cobra.(*Command).Execute
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
sigs.k8s.io/krew/cmd/krew/cmd.Execute
	/home/runner/work/krew/krew/cmd/krew/cmd/root.go:75
main.main
	/home/runner/work/krew/krew/cmd/krew/main.go:25
runtime.main
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/proc.go:225
runtime.goexit
	/opt/hostedtoolcache/go/1.16.0/x64/src/runtime/asm_amd64.s:1371
```

Darwin Obinnas-MacBook-Pro.local 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64, Catalina

**Comment by ahmetb** [comment]
Do you think this tool would be used within kubectl? 

This feels more like kubeadm, which is better off as a standalone tool as well. If this is along the same lines, I recommend not making it a plugin. Plugins are typically used for working with the resources inside the cluster.

I will take a look at why tests are failing.

**Comment by ahmetb** [comment]
> I will take a look at why tests are failing.

I can't repro the test failures locally with krew v0.4.2 (it looks like the zip extraction bug we've fixed in this version, so I am not sure why the CI is not picking up on that).

> [regarding naming]

We typically ask plugin developers to not name their plugins with kube* prefix. That is another indicator that this tool might be developed as and more suitable as a standalone tool rather than a kubectl plugin.

**Comment by xmudrii** [comment]
@ahmetb Thanks for the feedback. I'll try to answer your questions and provide my 2 cents.

> This feels more like kubeadm, which is better off as a standalone tool as well. If this is along the same lines, I recommend not making it a plugin.

It's a cluster management tool, but it works completely differently than kubeadm. Kubeadm requires you to create the instances, configure those instances, install the packages, run Kubeadm on all nodes, then deploy CNI and what else you need. There are a lot of manual steps involved 

KubeOne requires you only to create instances and even that is easier because we provide example Terraform configs. Everything else — provisioning, upgrading, repairing, etc... — is done by modifying the configuration file and running a single command (`kubeone apply`). KubeOne actually uses Kubeadm under the hood, but it abstracts away a lot of complexity and tasks.

As a reference and if you're interested, the following two docs describe installing and upgrading use cases:

* [Creating clusters](https://docs.kubermatic.com/kubeone/v1.3/tutorials/creating_clusters/)
* [Upgrading clusters](https://docs.kubermatic.com/kubeone/v1.3/tutorials/upgrading_clusters/)

If I would make a parallel to some existing tool, I would compare it to Kind. Kind also uses kubeadm under the hood, although Kind works only with containers and doesn't have day 2 operations (upgrading, repairing, etc). Unfortunately, Kind is not a kubectl plugin, so it's maybe not the greatest example. :upside_down_face: 

> Plugins are typically used for working with the resources inside the cluster. (A plugin in the same situation is #1636.)

> Do you think it makes sense that this tool would be used within kubectl?

While KubeOne doesn't work with Kubernetes resources, it works with Kubernetes clusters. Having kubectl manage both your Kubernetes resources and your clusters can be quite powerful, and I can definitely see users using it this way.

I hope this is not too long and I hope this is going to be at least a little bit helpful to make the decision. :slightly_smiling_face:

**Comment by ahmetb** [comment]
> Having kubectl manage both your Kubernetes resources and your clusters can be quite powerful, and I can definitely see users using it this way.

I think we might have a disagreement about this. :) It's definitely a gray area that we keep running into –whether something should be a kubectl plugin or just a standalone tool.

Think of it this way: Do you think `kubectl` would have subcommands that are used to create or upgrade clusters? The answer is probably a no, as kubectl sticks to managing the resources within an existing cluster. That's why I am inclined towards rejecting plugins of this type.

/kind gray-area

**Comment by xmudrii** [comment]
> I think we might have a disagreement about this. :) It's definitely a gray area that we keep running into –whether something should be a kubectl plugin or just a standalone tool.

I agree this is a gray area. It’s never easy to make such decisions.

> Think of it this way: Do you think kubectl would have subcommands that are used to create or upgrade clusters? The answer is probably a no, as kubectl sticks to managing the resources within an existing cluster. That's why I am inclined towards rejecting plugins of this type.

Looking from that standpoint, I definitely agree, but that brings the point could plugins bring more features that were not supposed to be part of the core. IMO, to some degree it would be useful, but on the other side, it makes it really hard to draw the line what should go in and what shouldn’t.

Having the line at what manages resources makes a lot of sense to me and what’s important — it makes the decision much easier.

I think we mostly agree about this one, so I’m going to trust your judgment on how to continue with this. 🙂

**Comment by ahmetb** [comment]
The more I read the documentation for KubeOne, more I think that this tool _structurally_ feels like kubeadm, minikube, kind etc as you said. It has its own "apply" command etc, too. So my recommendation would be to keep it as a standalone tool. If I were a user, I would not expect to see kubeone as a subcommand.

If you insist on making it a kubectl plugin, you can still distribute using the custom indexes feature: https://krew.sigs.k8s.io/docs/developer-guide/custom-indexes/ That way, users can still install the plugin via Krew, and they'll be downloading the plugin manifest from your repository instead of the central krew-index.

Thanks for the discussion here, until next time.
/close

**Pull request closed without merging**